File: java/src/0024.java
import static org.assertj.core.api.Assertions.assertThat;
import static structs.util.LinkedList.of;

import java.time.Instant;
import java.time.temporal.ChronoUnit;

import structs.linkedlist.ListNode;

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode swapPairs(ListNode head) {
        var fake = new ListNode(0, head);
        var curr = fake;
        while (curr.next != null && curr.next.next != null) {
            var next = curr.next;
            var temp = next.next.next;
            curr.next = next.next;
            next.next.next = next;
            next.next = temp;
            curr = next;
        }
        return fake.next;
    }

    public static void main(String[] args) {
        var o = new Solution();

        assertThat(o.swapPairs(of(1, 2, 3, 4))).isEqualTo(of(2, 1, 4, 3));
        assertThat(o.swapPairs(null)).isNull();
        assertThat(o.swapPairs(of(1))).isEqualTo(of(1));

        var start = Instant.now();
        for (var i = 0; i < 100; i++) {
            o.swapPairs(of(1, 2, 3, 4, 5));
            o.swapPairs(of(1));
            o.swapPairs(of(1, 2));
        }
        System.out.println(ChronoUnit.MICROS.between(start, Instant.now()));
    }
}

File: cpp/src/0024.cpp
/*
[24. Swap Nodes in Pairs (Medium)](https://leetcode.com/problems/swap-nodes-in-pairs/)

@topic: Linked List
@topic: Recursion

Constraints:
  - The number of nodes in the&nbsp;list&nbsp;is in the range `[0, 100]`.
  - `0 <= Node.val <= 100`
*/

#include "structs/LinkedList.hpp"

class Solution {
   public:
    ListNode* swapPairs(ListNode* head) {
        auto fake = ListNode(0, head);
        auto curr = &fake;
        while (curr->next && curr->next->next) {
            auto next = curr->next;
            head = next->next->next;
            curr->next = next->next;
            next->next->next = next;
            next->next = head;
            curr = next;
        }
        return fake.next;
    }
};

#include <cassert>
#include <chrono>
#include <iostream>
#include <vector>

using std::tuple;
using std::vector;
using std::chrono::duration_cast;
using std::chrono::microseconds;
using std::chrono::system_clock;

int main() {
    using namespace LinkedList;

    Solution o;

    vector<tuple<ListNode*, ListNode*>> CASES = {
        {of({1, 2, 3, 4}), of({2, 1, 4, 3})},
        {of({}), of({})},
        {of({1}), of({1})},
    };

    for (auto& [head, excepted] : CASES) {
        assert(eq(o.swapPairs(head), excepted));
    }

    auto start = system_clock::now();
    for (auto i = 0; i < 100000; ++i) {
        o.swapPairs(of({1, 2, 3, 4}));
        o.swapPairs(of({}));
        o.swapPairs(of({1}));
    }
    auto end = system_clock::now();
    std::cout << duration_cast<microseconds>(end - start).count() << std::endl;
}
